test/opt/test_def_use.cpp

```diff
@@ -0,0 +1,599 @@
+            {1, "%1 = OpTypeBool"},
+            {2, "%2 = OpTypeVector %10 2"},
+            {3, "%3 = OpTypeVector %10 3"},
+            {4, "%4 = OpTypeVector %10 4"},
+            {5, "%5 = OpTypeMatrix %20 2"},
+            {6, "%6 = OpTypeMatrix %3 3"},
+            {7, "%7 = OpTypeMatrix %40 4"},
+          },
+          {
+            {10,
+              {
+                "%2 = OpTypeVector %10 2",
+                "%3 = OpTypeVector %10 3",
+                "%4 = OpTypeVector %10 4",
+              }
+            },
+            {20, {"%5 = OpTypeMatrix %20 2"}},
+            {3, {"%6 = OpTypeMatrix %3 3"}},
+            {40, {"%7 = OpTypeMatrix %40 4"}},
+          },
+        },
+      },
+    })
+);
+// clang-format on
+
+struct KillDefCase {
+  const char* before;
+  std::vector<uint32_t> ids_to_kill;
+  const char* after;
+  InstDefUse du;
+};
+
+using KillDefTest = ::testing::TestWithParam<KillDefCase>;
+
+TEST_P(KillDefTest, Case) {
+  const auto& tc = GetParam();
+
+  // Build module.
+  const std::vector<const char*> text = {tc.before};
+  std::unique_ptr<ir::Module> module =
+      SpvTools(SPV_ENV_UNIVERSAL_1_1).BuildModule(JoinAllInsts(text));
+  ASSERT_NE(nullptr, module);
+
+  // Analyze def and use.
+  opt::analysis::DefUseManager manager;
+  manager.AnalyzeDefUse(module.get());
+
+  // Do the substitution.
+  for (const auto id : tc.ids_to_kill) manager.KillDef(id);
+
+  EXPECT_EQ(tc.after, DisassembleModule(module.get()));
+  CheckDefUse(tc.du, manager.id_to_defs(), manager.id_to_uses());
+}
+
+// clang-format off
+INSTANTIATE_TEST_CASE_P(
+    TestCase, KillDefTest,
+    ::testing::ValuesIn(std::vector<KillDefCase>{
+      { // no def, no use, no kill
+        "", {}, "", {}
+      },
+      { // kill nothing
+        "%1 = OpTypeBool "
+        "%2 = OpTypeVector %1 2 "
+        "%3 = OpTypeVector %1 3 ",
+        {},
+        "%1 = OpTypeBool\n"
+        "%2 = OpTypeVector %1 2\n"
+        "%3 = OpTypeVector %1 3",
+        {
+          {
+            {1, "%1 = OpTypeBool"},
+            {2, "%2 = OpTypeVector %1 2"},
+            {3, "%3 = OpTypeVector %1 3"},
+          },
+          {
+            {1,
+              {
+                "%2 = OpTypeVector %1 2",
+                "%3 = OpTypeVector %1 3",
+              }
+            },
+          },
+        },
+      },
+      { // kill id used, kill id not used, kill id not defined
+        "%1 = OpTypeBool "
+        "%2 = OpTypeVector %1 2 "
+        "%3 = OpTypeVector %1 3 "
+        "%4 = OpTypeVector %1 4 "
+        "%5 = OpTypeMatrix %3 3 ",
+        {1, 3, 5, 10},
+        "OpNop\n"
+        "%2 = OpTypeVector %1 2\n"
+        "OpNop\n"
+        "%4 = OpTypeVector %1 4\n"
+        "OpNop",
+        {
+          {
+            {2, "%2 = OpTypeVector %1 2"},
+            {4, "%4 = OpTypeVector %1 4"},
+          },
+          {
+            {1,
+              {
+                "%2 = OpTypeVector %1 2",
+                "OpNop", // %3 = OpTypeVector %1 3
```

Comment:
This is unexpected.  The OpNop is no longer a use of 1. 
KillDef should remove the uses , I think.